### PR TITLE
Add a retry mechanism to obtain the lock for inserting products through the API

### LINKED ISSUE #51412: The SKU (xxx) you are trying to insert is already under processing [woocommerce_rest_product_not_created] error
### Prerequisites

- [ ] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

We run over 100 webshops on woo. All shops are connected with ERP software for catalog (products, categories, attributes, brands etc etc). We use the Woo REST API for communication.

Last weeks we run into a lot of problems where when we try to update a product we get this error:
The SKU (xxx) you are trying to insert is already under processing [woocommerce_rest_product_not_created] error

This is the json we send:
`{
  "attributes": [
    {
      "id": "4",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "5",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "6",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "7",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "8",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "2",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    },
    {
      "id": "14",
      "options": [
        ""
      ],
      "visible": true,
      "option": "",
      "variation": false
    }
  ],
  "brands": [
    "646"
  ],
  "catalog_visibility": "visible",
  "categories": [],
  "description": "Handcreme kokosnoot 75 ml Urtekram",
  "meta_data": [
    {
      "key": "_product_field_ingredienten",
      "value": ""
    },
    {
      "key": "_product_field_Gebruiksaanwijzing",
      "value": ""
    },
    {
      "key": "_woocommerce_gpf_data",
      "value": {
        "brand": "Urtekram",
        "gtin": "5765228837894",
        "mpn": "15217",
        "google_product_category": "6261"
      }
    },
    {
      "key": "_product_field_Samenstelling",
      "value": ""
    },
    {
      "key": "_product_field_barcodeuitmplus",
      "value": "5765228837894"
    },
    {
      "key": "mplus_certificaat",
      "value": ""
    },
    {
      "key": "mplus_subtitel",
      "value": "Laat de natuur je een handje helpen!"
    }
  ],
  "name": "Handcreme kokosnoot 75 ml Urtekram",
  "regular_price": "6.6",
  "shipping_class": "pakket",
  "short_description": "Handcreme kokosnoot",
  "sku": "957957",
  "status": "publish",
  "stock_quantity": 1,
  "stock_status": "instock",
  "tax_class": "standard",
  "tax_status": "taxable",
  "type": "simple",
  "weight": "1"
}`

After searching arround we found that when you request for a product that does no longer excists in the interface of Woo it might be that it does exist in the: wp_postmeta and wp_wc_product_meta_lookup tables. We've cleaned those from orphaned SKU's and than we can access the endpoint again and create the products. It seems that earlier the product may have existed, was trashed but the postmeta an meta lookup was not cleaned properly?

I am sharing this because I couldn find any hits on the error:   The SKU (xxx) you are trying to insert is already under processing [woocommerce_rest_product_not_created] error

Maybe others also runned or will run into this and then they might find this helpfull to read.

### Expected behavior

I would expect the Woo rest API to be able to check of a SKU is truly existing and not orphaned data in some table.

### Actual behavior

The Woo rest API seems to use the postmeta or meta_lookup table (due to performance issues no INNER JOIN with the actual post i guess)

### Steps to reproduce

I don't know what caused the SKU to get orphaned, the trash was empty when we looked so we couldn't find when the product were deleted, could be years ago.

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.2.3
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.8.1
Log Directory Writable: ✔
WP Version: ❌ 6.5.5 - There is a newer version of WordPress available (6.6.2)
WP Multisite: ✔
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: –
Language: en_GB
External object cache: ✔

### Server Environment ###

Server Info: nginx/1.18.0
PHP Version: 8.3.8
PHP Post Max Size: 1 GB
PHP Time Limit: 300
PHP Max Input Vars: 5000
cURL Version: 7.81.0
OpenSSL/3.0.2

SUHOSIN Installed: –
MySQL Version: 8.0.39-0ubuntu0.22.04.1
Max Upload Size: 5 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

acf-field: 2
acf-field-group: 1
attachment: 29942
customize_changeset: 1
fgf_master_log: 34
fgf_rules: 4
fusion_icons: 1
fusion_tb_layout: 1
fusion_tb_section: 1
mc4wp-form: 1
nav_menu_item: 66
oembed_cache: 19
page: 39
post: 4
product: 7214
product_variation: 8
revision: 1182
shop_coupon: 8
shop_order: 561
shop_order_placehold: 1376
shop_order_refund: 106
wc_order_email: 3
wc_order_status: 14
wc_reminder_email: 15
wpcf7_contact_form: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (101) ###

Avada Custom Checkout: by Ivar Zuidhoek for MplusWebshops.nl – 1.2.5
WooCommerce Clone Orders: by Vibe Agency – 1.5.7 (update to version 1.5.9 is available)
Antispam Bee: by pluginkollektiv – 2.11.6
CardGate: by CardGate – 3.1.28
Classic Widgets: by WordPress Contributors – 0.3
Cloudflare: by Cloudflare
Inc. – 4.12.7

Contact Form 7: by Takayuki Miyoshi – 5.9.6
Agy: by Patrick Posner – 4.3.5
Disable Bloat for WordPress & WooCommerce: by Disable Bloat – 3.4.9
Yoast Duplicate Post: by Enrico Battocchi & Team Yoast – 4.5
GTM4WP - A Google Tag Manager (GTM) plugin for WordPress: by Thomas Geiger – 1.20.2
FacetWP: by FacetWP
LLC – 4.3

Free Gifts for WooCommerce: by Flintop – 10.4.0
Avada Builder: by ThemeFusion – 3.11.9
Avada Core: by ThemeFusion – 5.11.9
Giftable for WooCommerce: by Decom – 1.0.5
Loco Translate: by Tim Whitlock – 2.6.10
Login Security Captcha: by ScriptsTown – 1.6.6
Mailgun: by Mailgun – 1.9.9
Members: by MemberPress – 3.2.9
Mplus koppeling: by MplusWEBSHOPS – 3.9.8
Permalink Manager Pro: by Maciej Bis – 2.4.3.5
Redirection: by John Godley – 5.4.2
Say What?: by Ademti Software – 2.2.3
FastPicker: by  – 1.0.2
WooCommerce Smart Reminder Emails: by WP Overnight – 2.7.3
Variation Swatches for WooCommerce: by Emran Ahmed – 2.1.1
WooCommerce Brands: by WooCommerce – 1.6.54 (update to version 1.7.6 is available)
FOX - Currency Switcher Professional for WooCommerce: by realmag777 – 1.4.2.1
WooCommerce Dynamic Pricing: by Element Stark – 3.2.3 (update to version 3.2.11 is available)
WooCommerce EU VAT Number: by WooCommerce – 2.8.8 (update to version 2.9.7 is available)
Woocommerce Gift Manager: by SolverCircle – 2.6
MyParcel: by MyParcel – 4.23.0
WooCommerce Order Status Manager: by SkyVerge – 1.15.2 (update to version 1.15.3 is available)
Woocommerce Partial Orders Pro: by Code Ninjas – 1.9.3
WooCommerce Points and Rewards: by WooCommerce – 1.7.42 (update to version 1.8.8 is available)
WooCommerce Google Product Feed: by Ademti Software Ltd. – 10.11.0 (update to version 11.0.13 is available)
WooCommerce: by Automattic – 9.2.3 (update to version 9.3.1 is available)
Yuluma Google Translate: by  – 0.0.1
WordPress Importer: by wordpressdotorg – 0.8.2
Yoast SEO: by Team Yoast – 23.4
WP All Import Pro: by Soflyy – 4.8.0
WP Crontrol: by John Blackbourn – 1.16.3
WP All Import - User Import Add-On Pro: by Soflyy – 1.1.8
WP All Import - WooCommerce Import Add-On Pro: by Soflyy – 3.3.6
Yuluma Address Autocomplete: by  – 0.0.1
Yuluma Automation: by Yuluma – 1.2.91
Disable Bloat for WordPress & WooCommerce: by Disable Bloat – 3.4.9
FiboSearch - AJAX Search for WooCommerce (Pro): by FiboSearch Team – 1.28.0
Agy: by Patrick Posner – 4.3.5
Antispam Bee: by pluginkollektiv – 2.11.6
Avada Builder: by ThemeFusion – 3.11.9
Avada Core: by ThemeFusion – 5.11.9
Avada Custom Checkout: by Ivar Zuidhoek for MplusWebshops.nl – 1.2.5
CardGate: by CardGate – 3.1.28
Classic Widgets: by WordPress Contributors – 0.3
Cloudflare: by Cloudflare
Inc. – 4.12.7

Contact Form 7: by Takayuki Miyoshi – 5.9.6
Giftable for WooCommerce: by Decom – 1.0.5
GTM4WP - A Google Tag Manager (GTM) plugin for WordPress: by Thomas Geiger – 1.20.2
Loco Translate: by Tim Whitlock – 2.6.10
Mailgun: by Mailgun – 1.9.9
Mplus koppeling: by MplusWEBSHOPS – 3.9.8
Permalink Manager Pro: by Maciej Bis – 2.4.3.5
WooCommerce Points and Rewards: by WooCommerce – 1.7.42 (update to version 1.8.8 is available)
Really Simple CAPTCHA: by Takayuki Miyoshi – 2.3
PDF Invoices & Packing Slips for WooCommerce: by WP Overnight – 3.8.4
reBusted!: by Simon Prosser – 1.1
Redirection: by John Godley – 5.4.2
Say What?: by Ademti Software – 2.2.3
Variation Swatches for WooCommerce: by Emran Ahmed – 2.1.1
WooCommerce: by Automattic – 9.2.3 (update to version 9.3.1 is available)
Woocommerce Gift Manager: by SolverCircle – 2.6
Woocommerce Partial Orders Pro: by Code Ninjas – 1.9.3
WooCommerce Smart Reminder Emails: by WP Overnight – 2.7.3
FOX - Currency Switcher Professional for WooCommerce: by realmag777 – 1.4.2.1
Yoast Duplicate Post: by Enrico Battocchi & Team Yoast – 4.5
Yoast SEO: by Team Yoast – 23.4
Yuluma Automation: by Yuluma – 1.2.91
Advanced Custom Fields PRO: by WP Engine – 6.3.3
Better Search Replace: by WP Engine – 1.4.7
PDF Invoices & Packing Slips for WooCommerce - Premium Templates: by WP Overnight – 2.21.8
WooCommerce Brands: by WooCommerce – 1.6.54 (update to version 1.7.6 is available)
WooCommerce Dynamic Pricing: by Element Stark – 3.2.3 (update to version 3.2.11 is available)
WooCommerce EU VAT Number: by WooCommerce – 2.8.8 (update to version 2.9.7 is available)
WooCommerce Google Product Feed: by Ademti Software Ltd. – 10.11.0 (update to version 11.0.13 is available)
WooCommerce Order Status Manager: by SkyVerge – 1.15.2 (update to version 1.15.3 is available)
Free Gifts for WooCommerce: by Flintop – 10.4.0
FunnelKit Cart for WooCommerce: by FunnelKit – 1.7.1
FastPicker: by  – 1.0.2
Yuluma Address Autocomplete: by  – 0.0.1
Laposta WooCommerce: by Laposta - Stijn van der Ree – 1.7.1
PDF Invoices & Packing Slips for WooCommerce - Professional: by WP Overnight – 2.15.9
WP Rocket: by WP Media – 3.16.2.1
WP Crontrol: by John Blackbourn – 1.16.3
FiboFilters: by FiboFilters Team – 1.2.0
WP Umbrella: by WP Umbrella - Backup & Manage WordPress – 2.15.4
Freemius Fixer: by Freemius – 1.0.0
Yuluma Plugin Updater: by  – 0.2.1
Redis Object Cache: by Till Krüss – 2.5.2
Migrate Guru: by Migrate Guru – 5.65

### Inactive Plugins (4) ###

KDZ: by Chris Kanger – 0.0.15
WooCommerce PDF Invoices & Packing Slips Professional: by Ewout Fernhout – 2.9.2
WP All Export - WooCommerce Export Add-On Pro: by Soflyy – 1.0.6
WP All Export Pro: by Soflyy – 1.8.3

### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php
maintenance.php: maintenance.php
object-cache.php: Redis Object Cache Drop-In

### Must Use Plugins (1) ###

WP Umbrella: by WP Umbrella – 1.0.0

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: dgfw_gift (dgfw_gift)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 155 KB

### WC Pages ###

Shop base: #7 - /winkel/
Cart: #8 - /winkelmand/ -  Contains the [woocommerce_cart] shortcode
Checkout: #9 - /afrekenen/ -  Contains the [woocommerce_checkout] shortcode
My account: #10 - /mijn-account/
Terms and conditions: #19686 - /algemene-voorwaarden/

### Theme ###

Name: Avada Child
Version: 1.0.0
Author URL: https://theme-fusion.com
Child Theme: ✔
Parent Theme Name: Avada
Parent Theme Version: 7.11.9
Parent Theme Author URL: https://themeforest.net/user/ThemeFusion
WooCommerce Support: ✔

### Templates ###

Overrides: Avada/woocommerce/cart/cart.php
Avada/woocommerce/checkout/form-pay.php
Avada/woocommerce/checkout/review-order.php
Avada/woocommerce/single-product/short-description.php
Avada/woocommerce/single-product/tabs/additional-information.php
Avada/woocommerce/single-product/tabs/description.php


### WooCommerce Google Product Feed feeds ###

google: Google merchant centre product feed
  Type: google
  Category filter: -
  Categories: -

### WooCommerce Google Product Feed options ###

Include variations in feed: Enabled
Send "item group ID": -
Expanded schema markup: -
Debug key: 9eb7ed81-e662-4697-a4eb-ff785bc81fc3

### WooCommerce Google Product Feed fields ###

Availability (backordered products): Defaults to "in stock".
Availability (in stock products): Defaults to "in stock".
Availability (out of stock products): Defaults to "out of stock".
Brand: Pre-populates from product_brand taxonomy.
Condition: Defaults to "new".
Delivery label: Pre-populates from product_brand taxonomy.
Global Trade Item Number (GTIN): Pre-populates from product sku.
Google Product Category: 
Manufacturer Part Number (MPN): Pre-populates from product sku.
Product Type: 
Product description: Main product description only (short preferred)
Size system: 
Title: Pre-populates from product product_title.
Adult content: Defaults to "no".
Custom label 0: Defaults to "shipping_label". Pre-populates from product_shipping_class meta field.
Energy label image link: 

### WooCommerce Google Product Feed DB status ###

Database version: 16
Active database version: 16
wc_gpf_render_cache: Empty
woocommerce_gpf_google_taxonomy (en-US): 5595 items
woocommerce_gpf_google_taxonomy (nl-NL): 5595 items

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: experimental-blocks
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system

Daily Cron: ✔ Next scheduled: 2024-09-16 13:21:04 +02:00
Options: ✔
Notes: 13
Onboarding: completed

### Action Scheduler ###

Complete: 1,812
Oldest: 2024-09-09 15:04:01 +0200
Newest: 2024-09-16 14:46:52 +0200

Failed: 21
Oldest: 2022-11-16 15:07:02 +0100
Newest: 2024-07-11 09:48:29 +0200

Pending: 3
Oldest: 2024-09-16 18:06:02 +0200
Newest: 2024-09-19 03:00:00 +0200


### Status report information ###

Generated at: 2024-09-16 14:46:55 +02:00
```
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by albarin**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a retry mechanism for the query to obtain the lock when inserting products through the API.
The insert query can potentially end up in a deadlock if there is a high concurrency of requests, resulting in a false negative for the SKU lock and, consequently, products not being created when they should.
https://github.com/woocommerce/woocommerce/blob/fc2ee61bf86ab77494430faef7ae0692cbcd9db1/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L124-L132

This was causing some users to get the `The SKU (XXX) you are trying to insert is already under processing` error when the SKU was unique and inserted for the first time.

Closes https://github.com/woocommerce/woocommerce/issues/51412

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**On trunk (to reproduce the error)**
1. Add the following lines in your `wp-config.php` file (if you didn't have them already).
```php
define( 'WP_DEBUG', true );
define( 'WP_DEBUG_LOG', true );
```
2. Copy the following script to your local
```bash
#!/bin/bash
for i in {1..10}
do
  curl --insecure -X POST [YOUR_URL]/wp-json/wc/v3/products \
    -u [YOUR_CONSUMER_KEY]:[YOUR_CONSUMER_SECRET] \
    -H "Content-Type: application/json" \
    -d "{
      \"name\": \"Premium Quality\",
      \"type\": \"simple\",
      \"sku\": \"saidofji-$i\",
      \"regular_price\": \"21.99\",
      \"description\": \"Pellentesque\"
    }" &
done
wait
```
4. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=keys` and create a new key with `Read/Write` permissions and copy and paste the key and secret into the previous script.
5. Replace the URL in the script with your store URL.
6. Run the script locally until you get an error similar to this: `The SKU (XXX) you are trying to insert is already under processing`. You may need to run the script multiple times until it happens, make sure to change the SKU in the script in each execution (or **delete permanently** the products from your store).
7. When it happens, go to the `debug.log` and there should be an error similar to: `WordPress database error Deadlock found when trying to get lock; try restarting transaction for query INSERT INTO...` and some of the products won't be present in `/wp-admin/edit.php?post_type=product`.

**On this branch**
1. Run the script again multiple times and make sure in every execution you get the 10 products created on `/wp-admin/edit.php?post_type=product`.
2. If you go to the `debug.log` you may encounter the `deadlock` error, but the product should have been created anyway because of the retries.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add a retry mechanism to the query to obtain the SKU lock for product creation through the API.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 1]
```diff
@@ -145,8 +145,17 @@ private function obtain_lock_on_sku_for_concurrent_requests( $product ) {
...
 		}
 
-		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
-		$result = $wpdb->query( $query );
+		// The insert query can potentially result in a deadlock if there is high concurrency
+		// when trying to insert products, which will result in a false negative for SKU lock
+		// and incorrectly products not being created.
+		// To mitigate this, we will retry the query 3 times before giving up.
+		for ( $attempts = 0; $attempts < 3; $attempts++ ) {
+			// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
Comment (barryhughes) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:157]: When `$attempts > 0` should we introduce a very short delay or, if not, should we provide the opportunity for deployments to do so via an action hook?

I'm not sure this is necessary—we could definitely start here and see what further feedback comes in—but wanted to pause and ask.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 2]
```diff
@@ -145,8 +145,17 @@ private function obtain_lock_on_sku_for_concurrent_requests( $product ) {
...
+			$result = $wpdb->query( $query );
+			if ( false !== $result ) {
+				break;
+			}
```
Comment (barryhughes) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:161]: If `$result === false` should we log something?

Comment (barryhughes) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:161]: ... And/or should we do so if `$result === false && $attemps === 2` (ie, if we are at the end of our allowed retries)?

Comment (albarin) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:161]: Good point!
What do you think about adding something like this after the loop:
```php
if ($result === false) {
	wc_get_logger()->warning(
		sprintf(
			'Failed to obtain SKU lock for product: ID "%d" with SKU "%s" after %d attempts.',
			$product_id,
			$sku,
			$attempts,
		),
		array(
			'error' => $wpdb->last_error,
		)
	);
}
```

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 1 above)
Comment (albarin) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:157]: What about we start adding that short delay after the first attempt and see how it goes:
```php
for ( $attempts = 0; $attempts < 3; $attempts++ ) {
	if ( $attempts > 1 ) {
		usleep( 10000 );
	}

	// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
	$result = $wpdb->query( $query );
	if ( false !== $result ) {
		break;
	}
}
```
And later, if necessary, we can think about adding a hook? 🤔

Comment (barryhughes) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:157]: That sounds great, and 10ms is probably a pretty reasonable default.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php (hunk 2 above)
Comment (barryhughes) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:161]: Perfect!

**Comment by albarin** [comment]
Thanks for you review @barryhughes, I've addressed your feedback here: 1f097bd849fe9ab6821c7b928fdab28c3e6d5702

**Comment by barryhughes** [comment]
Just noting I found the following modification (for the test snippet) to be useful, especially when repeating the test:

```sh
#!/bin/bash

random_string=$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 8 | head -n 1)

for i in {1..4}
do
  curl --insecure -X POST SITE_ADDRESS/wp-json/wc/v3/products \
    -u [YOUR_CONSUMER_KEY]:[YOUR_CONSUMER_SECRET] \
    -H "Content-Type: application/json" \
    -d "{
      \"name\": \"Premium Quality\",
      \"type\": \"simple\",
      \"sku\": \"$random_string\",
      \"regular_price\": \"21.99\",
      \"description\": \"Pellentesque\"
    }" &
done
wait
```

Just because I could re-run and poke at different things, without manually updating the SKU between runs. I also removed the `-$i` incrementor, so that near-parallel requests referenced the same SKU *(with* the incrementor, each request referenced a different SKU), and had more luck replicating the core problem that way.

File: plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php [hunk 3]
```diff
@@ -145,8 +145,35 @@ private function obtain_lock_on_sku_for_concurrent_requests( $product ) {
...
+
+			// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+			$result = $wpdb->query( $query );
+			if ( false !== $result ) {
+				break;
+			}
+		}
+
+		if ( false === $result ) {
+			wc_get_logger()->warning(
```
Comment (barryhughes) [inline plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:165]: I think this is still useful and we should keep it, but noting we also get related error logging 'for free', in the context of REST API requests:

Unable to obtain lock

Comment (barryhughes) [reply plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php:165]: (Apologies as I didn't notice that earlier ... but I still think our additional logging is valid.)

**Review by barryhughes — APPROVED** [review]
Looks great!

I was able to simulate the problem, and see our new logging in action. One testing note (including for completeness) is that we don't always seem to get an actual DB error to share in the log context, see the following screenshot:

Screenshot 2024-10-02 at 07-56-23 WooCommerce status ‹ WordPress Lab — WordPress

But we *could* capture something here, of course.

**Comment by PurpleWebCreative** [comment]
Do we know when the next version is due to be released please?

**Comment by barryhughes** [comment]
The next release (9.4) should be on 2024-11-04, but I'm guessing you're interested in *this* change which should land in 9.5 (2024-12-09). Sometimes the release plan changes, so you may wish to take a look at this [recent developer blog post](https://developer.woocommerce.com/2024/10/02/new-public-release-calendar/).

**Comment by nigeljamesstevenson** [comment]
Noting that this PR will now be included in the 9.4 release

**Pull request merged**